rdb/db.go

```diff
@@ -0,0 +1,245 @@
+//go:build rocksdb
+// +build rocksdb
+
+package rdb
+
+// #include <stdlib.h>
+// #include "rocksdb/c.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"unsafe"
+
+	tmdb "github.com/line/tm-db/v2"
+	"github.com/line/tm-db/v2/internal/util"
+)
+
+type RDB struct {
+	name   string
+	fn     string
+	db     *C.rocksdb_t
+	opts   *C.rocksdb_options_t
+	ropts  *C.rocksdb_readoptions_t
+	wopts  *C.rocksdb_writeoptions_t
+	wsopts *C.rocksdb_writeoptions_t
+	wlpopts *C.rocksdb_writeoptions_t
+}
+
+type rdbIterator struct {
+	db         *RDB
+	it         *C.rocksdb_iterator_t
+	ropts      *C.rocksdb_readoptions_t
+	reverse    bool
+	lowerBound []byte
+	upperBound []byte
+}
+
+func cerror(cerr *C.char) error {
+	if cerr == nil {
+		return nil
+	}
+	err := errors.New(C.GoString(cerr))
+	C.free(unsafe.Pointer(cerr))
+	return err
+}
+
+func b2c(b []byte) *C.char {
+	if len(b) == 0 {
+		return nil
+	}
+	return (*C.char)(unsafe.Pointer(&b[0]))
+}
+
+func NewDB(name string, dir string) (*RDB, error) {
+	var cerr *C.char
+
+	fn := filepath.Join(dir, name+".db")
+
+	bbto := C.rocksdb_block_based_options_create()
+	C.rocksdb_block_based_options_set_block_cache(bbto, C.rocksdb_cache_create_lru(C.size_t(1<<30)))
+	C.rocksdb_block_based_options_set_filter_policy(bbto, C.rocksdb_filterpolicy_create_bloom(C.int(10)))
+	defer C.rocksdb_block_based_options_destroy(bbto)
+
+	opts := C.rocksdb_options_create()
+	C.rocksdb_options_set_block_based_table_factory(opts, bbto)
+	C.rocksdb_options_set_create_if_missing(opts, 1)
+	C.rocksdb_options_increase_parallelism(opts, C.int(runtime.NumCPU()))
+	C.rocksdb_options_optimize_level_style_compaction(opts, 512*1024*1024)
+	C.rocksdb_options_optimize_level_style_compaction(opts, 512*1024*1024)
+	C.rocksdb_options_set_enable_pipelined_write(opts, 1)
+
+	ropts := C.rocksdb_readoptions_create()
+	wopts := C.rocksdb_writeoptions_create()
+	wsopts := C.rocksdb_writeoptions_create()
```

Comment:
I don't know what each opts does, could you add some comment if possible?

Reply:
Added comments inside struct definition.